components/form.js

```diff
@@ -44,67 +31,156 @@ export default class Form extends React.Component {
+        fetchBearerToken().then(data => {
+          if (!data || data.error) {
+            ("Error while getting the data ");
+            reject(data);
+            return;
+          }
+          this.setState({
+            token: { ...data, fetching_time: new Date().getTime() }
+          });
+          localStorage.setItem(
+            "Bearer Token",
+            JSON.stringify(this.state.token)
+          );
+          resolve(data);
+        });
+      } else {
+        this.setState({ token: token });
+        resolve(token);
+      }
+    });
   }
 
   getFormFields() {
-    // Fetch the data form fields from API, use it in state
-    let formFields = JSON.parse(localStorage.getItem("Form Fields"));
-    if (!(Array.isArray(formFields) && formFields.length)) {
-      fetchFormFields().then(data =>
+    this.getToken().then(res => {
+      // Fetch the data form fields from API, use it in state
+      let formFields = JSON.parse(localStorage.getItem("Form Fields"));
+      if (!formFields) {
+        formFields = [];
+      }
+      if (formFields.length === 0) {
+        fetchFormFields().then(data => {
+          this.setState({
+            formFields: data.results
+          });
+          localStorage.setItem(
+            "Form Fields",
+            JSON.stringify(this.state.formFields)
+          );
+        });
+      } else {
         this.setState({
-          formFields: data.results
-        })
+          formFields: formFields
+        });
+      }
+    });
+  }
+
+  isOnlineEvent = e => {
+    let failedPosts = [];
+
+    let formDataArray = JSON.parse(
+      localStorage.getItem("Failed Form Submission Data")
+    );
+
+    if (!formDataArray) {
+      formDataArray = [];
+    }
+
+    formDataArray.map(formData => {
+      postFormData(formData, this.state.token.access_token).then(response => {
+        if (response && !response.ok) {
+          failedPosts.push(formData);
+        }
+      });
+    });
+
+    if (failedPosts.length !== 0) {
+      localStorage.setItem(
+        "Failed Form Submission Data",
+        JSON.stringify(failedPosts)
       );
     } else {
-      this.setState({
-        formFields: formFields
-      });
+      localStorage.removeItem("Failed Form Submission Data");
     }
-  }
+  };
 
   componentDidMount() {
     this.getFormFields();
-    this.getToken();
   }
 
   componentDidUpdate(prevProps, prevState) {
-    localStorage.setItem("Form Fields", JSON.stringify(this.state.formFields));
-    localStorage.setItem("Bearer Token", JSON.stringify(this.state.token));
+    window.addEventListener("online", this.isOnlineEvent);
+  }
+
+  isObjEmpty(obj) {
+    return Object.entries(obj).length === 0 && obj.constructor === Object;
   }
 
   // Handler html For when the submit button is pressed
   onSubmit = e => {
     e.preventDefault();
-    postFormData(this.state.formData, this.state.token.access_token);
+    if (this.isObjEmpty(this.state.formData)) {
+      return;
```

Comment:
FYI here is where you'd warn the user about the input being empty :) - flagging just in case -